### PR TITLE
Add correlation id to Token Parameters and Authentication Result

### PR DESCRIPTION
**Pull request by shahzaibj**
**What does this PR achieve?**
In this PR, (at a high level) we've done the following:

- Accept (optional) correlation id from devs on `TokenParameters`
- Expose correlation id on `AuthenticationResult`
- Add tests for the above two points

**Why is this being done?**
This work was requested from CSS team and also customers have a desire to use their own correlation id. See this work item: https://dev.azure.com/IdentityDivision/Engineering/_workitems/edit/875728

Other libraries have also already implemented this. 

Here's an example of it in `DotNet`

- Correlation id on token params: [See this link](https://docs.microsoft.com/en-us/dotnet/api/microsoft.identity.client.abstractacquiretokenparameterbuilder-1.withcorrelationid?view=azure-dotnet#Microsoft_Identity_Client_AbstractAcquireTokenParameterBuilder_1_WithCorrelationId_System_Guid_)
- Correlation id on Authentication Result: [See this link](https://docs.microsoft.com/en-us/dotnet/api/microsoft.identity.client.authenticationresult.correlationid?view=azure-dotnet#Microsoft_Identity_Client_AuthenticationResult_CorrelationId)

Here's an example of it in `iOS`

- Correlation id on token params: [See this link](https://azuread.github.io/microsoft-authentication-library-for-objc/Classes/MSALTokenParameters.html#/c:objc(cs)MSALTokenParameters(py)correlationId)
- Correlation id on Authentication Result: [See this link](https://azuread.github.io/microsoft-authentication-library-for-objc/Classes/MSALResult.html#/c:objc(cs)MSALResult(py)correlationId)

**How are we doing it?**
We are following the same design as shown above in DotNet example which is to accept a `GUID` or `UUID` for correlation id on `TokenParameters` - this will be optional and can be specified via the builder.

If the dev does not specify a correlation id then we generate our own as we do today.

When returning the result (`Authentication Result`) back to the caller, we set the correlation id on this result - this is whatever correlation id was used for that request (which would be either the one we generated or the one passed in by the dev).

This PR also depends on common PR: https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/925

### CODE REVIEW

File: msal/src/main/java/com/microsoft/identity/client/TokenParameters.java [hunk 1]
```diff
@@ -137,6 +140,10 @@ public AccountRecord getAccountRecord() {
         return mAccountRecord;
     }
 
+    public String getCorrelationId() {
```
Comment (AdamBJohnsonx) [inline msal/src/main/java/com/microsoft/identity/client/TokenParameters.java:143]: Javadoc public methods.

File: msal/src/main/java/com/microsoft/identity/client/TokenParameters.java [hunk 2]
```diff
@@ -42,13 +43,15 @@
     private ClaimsRequest mClaimsRequest;
     private AccountRecord mAccountRecord;
     private AuthenticationScheme mAuthenticationScheme;
+    private String mCorrelationId;
```
Comment (AdamBJohnsonx) [inline msal/src/main/java/com/microsoft/identity/client/TokenParameters.java:46]: Is it possible that these could be final?

File: msal/src/main/java/com/microsoft/identity/client/AuthenticationResult.java [hunk 3]
```diff
@@ -108,4 +117,26 @@ public IAccount getAccount() {
...
+    @Override
+    public UUID getCorrelationId() {
+        return mCorrelationId;
+    }
+
+    private UUID sanitizeCorrelationId(@Nullable final String correlationId) {
+        final String methodName = "sanitizeCorrelationId";
+
+        if (TextUtils.isEmpty(correlationId)) {
+            Logger.warn(TAG + methodName, "Correlation was empty, returning null.");
```
Comment (shahzaibj) [inline msal/src/main/java/com/microsoft/identity/client/AuthenticationResult.java:131]: nit: missing id

File: msal/src/main/java/com/microsoft/identity/client/TokenParameters.java (hunk 2 above)
Comment (shahzaibj) [reply msal/src/main/java/com/microsoft/identity/client/TokenParameters.java:46]: Nope, these have setters and can get modified after initialization: https://github.com/AzureAD/microsoft-authentication-library-for-android/blob/dev/msal/src/main/java/com/microsoft/identity/client/TokenParameters.java#L80-L134

File: msal/src/main/java/com/microsoft/identity/client/TokenParameters.java (hunk 1 above)
Comment (shahzaibj) [reply msal/src/main/java/com/microsoft/identity/client/TokenParameters.java:149]: Addressed here: https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/1063/commits/6a71d1d9fc3d3329f4b608dd3702a39ff7099b62

File: msal/src/main/java/com/microsoft/identity/client/AuthenticationResult.java (hunk 3 above)
Comment (shahzaibj) [reply msal/src/main/java/com/microsoft/identity/client/AuthenticationResult.java]: Addressed here: https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/1063/commits/6a71d1d9fc3d3329f4b608dd3702a39ff7099b62

File: msal/src/main/java/com/microsoft/identity/client/TokenParameters.java [hunk 4]
```diff
@@ -201,6 +209,11 @@ public B withResource(final String resource) {
             return self();
         }
 
+        public B withCorrelationId(final UUID correlationId) {
```
[the remark this answers is not in the document]
Comment (shahzaibj) [reply msal/src/main/java/com/microsoft/identity/client/TokenParameters.java]: Addressed here: https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/1063/commits/6a71d1d9fc3d3329f4b608dd3702a39ff7099b62

File: msal/src/test/java/com/microsoft/identity/client/e2e/tests/mocked/AcquireTokenMockTest.java [hunk 5]
```diff
@@ -335,6 +342,136 @@ public void testAcquireTokenSilentFailureNoCallback() {
...
+                    }
+                })
+                .build();
+
+        mApplication.acquireToken(parameters);
+        flushScheduler();
+    }
+
+    @Test
+    public void testAcquireTokenSilentResultContainsProvidedCorrelationId() {
```
[the remark this answers is not in the document]
Comment (shahzaibj) [reply msal/src/test/java/com/microsoft/identity/client/e2e/tests/mocked/AcquireTokenMockTest.java:445]: @AdamBJohnsonx Added unit tests for `command parameters` here: https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/1063/commits/4b194eb18c69e64ab39006ff57bcdfb2dfe3c222

and added unit tests for `Authentication Result` here: https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/1063/commits/4b194eb18c69e64ab39006ff57bcdfb2dfe3c222

**Comment by shahzaibj** [comment]
@AdamBJohnsonx All of your comments have been addressed

File: msal/src/main/java/com/microsoft/identity/client/AuthenticationResult.java [hunk 6]
```diff
@@ -108,4 +117,26 @@ public IAccount getAccount() {
...
         return mAccessToken.getTarget().split("\\s");
     }
+
+    @Nullable
+    @Override
+    public UUID getCorrelationId() {
+        return mCorrelationId;
+    }
+
+    private UUID sanitizeCorrelationId(@Nullable final String correlationId) {
```
Comment (iambmelt) [inline msal/src/main/java/com/microsoft/identity/client/AuthenticationResult.java:127]: Could declare return value `@Nullable`

File: msal/src/test/java/com/microsoft/identity/client/AuthenticationResultTest.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
+package com.microsoft.identity.client;
```
Comment (iambmelt) [inline msal/src/test/java/com/microsoft/identity/client/AuthenticationResultTest.java:1]: License info?

File: msal/src/main/java/com/microsoft/identity/client/AuthenticationResult.java (hunk 6 above)
Comment (shahzaibj) [reply msal/src/main/java/com/microsoft/identity/client/AuthenticationResult.java:128]: Addressed here: https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/1063/commits/353f46523435797d9079cbf600a795d288eda4ed

File: msal/src/test/java/com/microsoft/identity/client/AuthenticationResultTest.java (hunk 7 above)
Comment (shahzaibj) [reply msal/src/test/java/com/microsoft/identity/client/AuthenticationResultTest.java:23]: Addressed here: https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/1063/commits/353f46523435797d9079cbf600a795d288eda4ed

**Pull request merged**